### PR TITLE
Support NNX ckpt in forward_pass_logit_checker

### PR DESCRIPTION
**Pull request by hengtaoguo**
# Description
Support NNX checkpoint in forward_pass_logit_checker.py:
- [New] Add a flag `--ckpt_type=["linen", "nnx"]` to support nnx ckpt. Default to linen.
- Implement `setup_decode_state_from_nnx()`, which now calls `from_pretrained()` to natively load an NNX model and shard properly across devices (instead of loading to CPU or duplicating across TPUs). This function will return a linen `maxtext_state` for decoding purposes.

When loading an NNX checkpoint that was saved via `TunixMaxTextAdapter` (e.g. an RL training checkpoint), the on-disk pytree can contain extra top-level keys such as to_nnx__rngs that are present in the checkpoint but absent from the target model's parameter tree. Passing both trees directly to `jax.tree.map` requires them to have identical structure, so the extra checkpoint keys cause a key-mismatch error. 
- `_filter_to_model_keys` prunes the restored checkpoint dict down to only the keys present in `model_arrays`, ensuring the two trees are structurally identical before the `jax.tree.map(_expand_checkpoint_to_model_shapes, ...)` call, without silently skipping any weights the model actually needs.

# Tests

Added unit test `test_returns_linen_train_state_and_annotations` which verifies that `setup_decode_state_from_nnx`:
1. Calls `from_pretrained` with `model_mode=MODEL_MODE_AUTOREGRESSIVE` (never touches a real checkpoint in CI)
2. Returns a linen `TrainState` with concrete `jax.Array` params nested under `state.params["params"]`
3. Extracts param values faithfully — each leaf is numerically identical to the source NNX model's `nnx.Param` values

End-to-end runs:
- NNX-RL with `base`, `--ckpt_type=nnx`: [example1](https://paste.googleplex.com/5791541516959744), [example2](https://paste.googleplex.com/5402751547867136)
- NNX-SFT without `base`, `--ckpt_type=nnx`: [example3](https://paste.googleplex.com/6319774511529984)
- Linen ckpt unimpacted: [example4](https://paste.googleplex.com/5626151453196288)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/utils/maxtext_utils.py [hunk 1]
```diff
@@ -1087,6 +1087,53 @@ def setup_decode_state(model, config, rng, mesh, checkpoint_manager):
...
+      filtered_tree = {k: v for k, v in tree.items() if "to_nnx__rngs" not in k}
+      if (
+          len(filtered_tree) == 1
+          and "value" in filtered_tree
+          and isinstance(filtered_tree["value"], (jax.Array, np.ndarray))
+      ):
+        return filtered_tree["value"]
+      return {k: unwrap_nnx_values(v) for k, v in filtered_tree.items()}
+    return tree
+
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply src/maxtext/utils/maxtext_utils.py]: Refactored to use `create_nnx_model` instead, which creates a properly sharded nnx model.

File: src/maxtext/utils/maxtext_utils.py [hunk 2]
```diff
@@ -1087,6 +1087,53 @@ def setup_decode_state(model, config, rng, mesh, checkpoint_manager):
...
+  _, state_mesh_annotations, state_mesh_shardings = get_abstract_state(model, None, config, rng, mesh, False)
+
+  loaded_params = load_orbax_checkpoint(config)
+  loaded_params = unwrap_nnx_values(loaded_params)
+  # NNX-RL checkpoints may have an extra 'base' nesting level, while NNX-SFT may not
+  raw_params = loaded_params["base"] if "base" in loaded_params else loaded_params
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply src/maxtext/utils/maxtext_utils.py]: Obsolete

File: src/maxtext/utils/maxtext_utils.py [hunk 3]
```diff
@@ -1087,6 +1087,52 @@ def setup_decode_state(model, config, rng, mesh, checkpoint_manager):
...
+
+  # Extract only Param values from the NNX model (filters out dropout/RNG state)
+  # nnx.state() returns a State object (custom pytree), which jax.device_put can't match
+  # against a plain-dict sharding spec. Convert to a plain nested dict first.
+  def _state_to_dict(tree):
+    if hasattr(tree, "items") and not isinstance(tree, jax.Array):
+      return {k: _state_to_dict(v) for k, v in tree.items()}
+    return tree
+
+  nnx_param_state = nnx.state(nnx_model, nnx.Param)
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply src/maxtext/utils/maxtext_utils.py]: Done

File: tests/utils/forward_pass_logit_checker.py [hunk 4]
```diff
@@ -425,7 +425,10 @@ def main(config, test_args):  # pylint: disable=W0621
     mesh = jax.sharding.Mesh(devices_array, config.mesh_axes)
     quant = quantizations.configure_quantization(config)
     maxtext_model = models.transformer_as_linen(config, mesh, quant=quant, model_mode=MODEL_MODE_TRAIN)
-    maxtext_state, _ = maxtext_utils.setup_decode_state(maxtext_model, config, rng1, mesh, None)
+    if test_args.ckpt_type == "linen":
```
Comment (entrpn) [inline tests/utils/forward_pass_logit_checker.py:451]: Can we use the already existing enable_nnx flag instead of a creating a new one ckpt_type?

File: src/maxtext/utils/maxtext_utils.py [hunk 5]
```diff
@@ -1087,6 +1087,54 @@ def setup_decode_state(model, config, rng, mesh, checkpoint_manager):
...
+    state: state with decode params loaded from the NNX checkpoint
+    state_mesh_annotations: the mesh annotations for the state
+  """
+  from flax import nnx  # pylint: disable=import-outside-toplevel
+  from maxtext.utils.model_creation_utils import create_nnx_model  # pylint: disable=import-outside-toplevel
+
+  _, state_mesh_annotations, state_mesh_shardings = get_abstract_state(model, None, config, rng, mesh, False)
+
+  # Load the NNX model with checkpoint (handles NNX, NNX-RL, and linen checkpoint formats)
+  nnx_model, _ = create_nnx_model(config, mesh=mesh, model_mode=MODEL_MODE_AUTOREGRESSIVE)
```
Comment (entrpn) [inline src/maxtext/utils/maxtext_utils.py]: why are we manually forcing sharding via device_put instead of letting the model creation utility do it using `with mesh, nn_partitioning.axis_rules(cfg.logical_axis_rules):`

File: src/maxtext/utils/maxtext_utils.py [hunk 6]
```diff
@@ -1087,6 +1087,54 @@ def setup_decode_state(model, config, rng, mesh, checkpoint_manager):
...
+    config: config object
+    rng: jax.prng key
+    mesh: jax.devices() mesh
+
+  Returns:
+    state: state with decode params loaded from the NNX checkpoint
+    state_mesh_annotations: the mesh annotations for the state
+  """
+  from flax import nnx  # pylint: disable=import-outside-toplevel
+  from maxtext.utils.model_creation_utils import create_nnx_model  # pylint: disable=import-outside-toplevel
```
Comment (shuningjin) [inline src/maxtext/utils/maxtext_utils.py]: nit: wondering why not import at top level?

**Review by shuningjin — APPROVED** [review]
Thanks, LGTM!

File: tests/utils/forward_pass_logit_checker.py (hunk 4 above)
Comment (shuningjin) [reply tests/utils/forward_pass_logit_checker.py:451]: My impression is that `enable_nnx=true`([src1](https://github.com/AI-Hypercomputer/maxtext/blob/3971206a2a8cedd052c930195aef4cccbe425b4e/src/maxtext/models/models.py#L247)) might be different from `create_nnx_model` ([src2](https://github.com/AI-Hypercomputer/maxtext/blob/3971206a2a8cedd052c930195aef4cccbe425b4e/src/maxtext/utils/model_creation_utils.py#L115)). Overriding the meaning might cause confusion.

File: src/maxtext/utils/maxtext_utils.py (hunk 5 above)
Comment (hengtaoguo) [reply src/maxtext/utils/maxtext_utils.py]: Thanks for the suggestion! Removed this `device_put` and now the function `from_pretrained` (formerly create_nnx_model) will handle the sharding by default.

File: src/maxtext/utils/maxtext_utils.py (hunk 6 above)
Comment (hengtaoguo) [reply src/maxtext/utils/maxtext_utils.py]: Thanks for pointing out! I feel these two dependencies are only used within the function `setup_decode_state_from_nnx`. And the AI suggests: 

> If `maxtext_utils` were to import model_creation_utils at file level in return, Python would hit a circular import error at module load time. The function-level import sidesteps this because it only runs when setup_decode_state_from_nnx is actually called, by which point both modules are fully initialized.

Comment (hengtaoguo) [reply src/maxtext/utils/maxtext_utils.py]: Actually although I could run this new feature on VM, g3 is complaining about cycle graph imports. So I moved this function into model_creation_utils (nnx related). Now all the imports are top-level, thanks!
